File: fel-adapter/src/de/proskor/fel/impl/ComponentFaultTreeImpl.scala
package de.proskor.fel.impl

import de.proskor.automation.impl.DiagramImpl
import de.proskor.automation.Diagram
import de.proskor.automation.Package
import de.proskor.automation.Repository
import de.proskor.fel.view.ArchitecturalView
import de.proskor.fel.view.ComponentFaultTree

class ComponentFaultTreeImpl(diagram: Diagram) extends ViewImpl(diagram) with ComponentFaultTree {
  private def fel: Package = {
    val repository = Repository.instance
    val model = repository.models.headOption getOrElse repository.models.add("Model", "Package")
    model.packages.find(_.name == "FEL") getOrElse model.packages.add("FEL", "Package")
  }
  
  override def getContext: ArchitecturalView = {
    val list = for {
      tv <- fel.element.taggedValues if tv.name == "context"
      (cft, context) = decode(tv.value)
      if cft == diagram
    } yield new ArchitecturalViewImpl(context)
    list.headOption orNull
  }

  private def decode(text: String): (Diagram, Diagram) = {
    val parts = text.split(":")
    val left = parts(0).toInt
    val right = parts(1).toInt
    val diagrams = allDiagrams(Repository.instance.models.head)
    (diagrams.find(_.id == left).get, diagrams.find(_.id == right).get) 
  }

  private def allDiagrams(pkg: Package): Seq[Diagram] =
    pkg.diagrams.toSeq ++ pkg.packages.flatMap(allDiagrams)

  override def setContext(context: ArchitecturalView) {
    val tv = fel.element.taggedValues.add("context", "TaggedValue")
    tv.value = diagram.id.toString + ":" + context.asInstanceOf[ArchitecturalViewImpl].diagram.id.toString
  }
}

File: ea-fel/src/de/proskor/fel/impl/ComponentFaultTreeImpl.java
package de.proskor.fel.impl;

import de.proskor.fel.view.ArchitecturalView;
import de.proskor.fel.view.ComponentFaultTree;
import de.proskor.model.Diagram;
import de.proskor.model.Package;
import de.proskor.model.Repository;
import de.proskor.model.TaggedValue;

public class ComponentFaultTreeImpl extends ViewImpl implements ComponentFaultTree {
	public ComponentFaultTreeImpl(Repository repository, Diagram peer) {
		super(repository, peer);
	}

	@Override
	public ArchitecturalView getContext() {
		final TaggedValue tv = this.getContextAssociation();
		if (tv != null) {
			final Repository repository = this.getRepository();
			final String[] values = tv.getValue().split(":isContextFor:");
			final Diagram context = repository.getDiagramById(Integer.valueOf(values[0]));
			return new ArchitecturalViewImpl(repository, context);
		}
		return null;
	}

	@Override
	public void setContext(ArchitecturalView view) {
		if (!(view instanceof ArchitecturalViewImpl))
			throw new IllegalArgumentException();

		final String value = ((ArchitecturalViewImpl) view).getPeer().getId() + ":isContextFor:" + this.getPeer().getId();
		final TaggedValue tv = this.getContextAssociation();
		if (tv != null) {
			tv.setValue(value);
		} else {
			final TaggedValue nt = this.getFEL().getElement().getTaggedValues().add("context", TaggedValue.TAGGEDVALUE);
			nt.setValue(value);
			
		}
	}

	private Package getFEL() {
		final Repository repository = this.getRepository();
		final Package model = repository.getModels().get(0);
		for (final Package pkg : model.getPackages()) {
			if (pkg.getName().equals("FEL"))
				return pkg;
		}
		return model.getPackages().add("FEL", Package.PACKAGE);
	}

	private TaggedValue getContextAssociation() {
		final Package fel = this.getFEL();
		final Repository repository = this.getRepository();
		final Diagram peer = this.getPeer();
		for (final TaggedValue tv : fel.getElement().getTaggedValues()) {
			if (tv.getName().equals("context")) {
				final String[] values = tv.getValue().split(":isContextFor:");
				final Diagram cft = repository.getDiagramById(Integer.valueOf(values[1]));
				if (peer.equals(cft))
					return tv;
			}
		}
		return null;
	}
}
